### ISSUE TITLE
[Proposal] Make dataInputSchema value with URI reference only

### ISSUE THREAD

**Issue by ricardozanini** [serverlessworkflow/specification#624]
**What would you like to be added**:

This proposal is to remove the possibility of having the `dataInputSchema` attribute in the workflow definition be a JSON Object. Instead, the `dataInputSchema` would be a URI only.

**Why is this needed**:

The reasoning behind this is to follow a pattern we already have in place for references on other standards like OpenAPI, gRPC, oData, Async, etc. Every time we refer to an external specification or standard is an external reference. This makes sense because we avoid mixing standards in our specification definitions.

Additionally, when composing the workflow on a YAML file, it's awkward defining a JSON Schema Object in a YAML structure.

**Comment by cdavernas** [comment]
@ricardozanini I totally disagree on that proposal. I actually created the PR to use an object. 

Not to speak about why I think external references should almost always be avoided, I think we can't and shouldn't restrict uses. Furthermore, not allowing the yes of the object would make the definition tightly coupled to an externally referenced file: kill the endpoint that hosts the schema, and you kill the workflow

Everywhere we allow external ref, we allow using an object directly, too. 

By no means should you write the schema using json when writing the definition in yaml, I'm not sure why you think that's needed: Json schema can be defined in both yaml or json.

**Comment by tsurdilo** [comment]
Currently its string or object type, so it would cover both. I think validation of workflow data input is very tied to the definition itself so its based on business requirements and not standards so defining it inline with the definition is ok imo.
Also having it inside the definition itself could help not having to do file lookups to define swagger-like docs from the definition too, wdyt?

Another thing is for testing i think being able to define the schema inline is super useful too.

Not sure i really see problem with inline defs either..but there could be something missing.

**Comment by ricardozanini** [comment]
> Everywhere we allow external ref, we allow using an object directly, too.

[It's not, that's the point](https://github.com/serverlessworkflow/specification/blob/main/schema/functions.json#L37-L40). It's always a String for a URI (openapi, async, graphql, etc.). All of them are external references. Or am I missing something? Adding an OpenAPI Spec in a function's  `operation` attribute is a no-go IMO. Where is the other place we can define gRPC, OpenAPI, and other standard references other than the `operation`, which is a URI?

> Furthermore, not allowing the use of an object would make the definition tightly coupled to an externally referenced file: kill the endpoint that hosts the schema, and you kill the workflow

Same for any other integrations we have with external standards.

A URI can be a local one (which will be in most cases). It shouldn't be an external reference all the time. This should be the recommended approach, IMHO. So you have a structure of:

- Workflow definitions
- Function definitions
- Events
- OpenAPIs
- Schemas
- Async
- GraphQL 
- etc

Adding all of them in a single file is tight coupling the whole workflow, which we should encourage to avoid.

Yes, I agree you can define an object in either YAML or JSON. However, this is a minor concern.

**Comment by cdavernas** [comment]
> [It's not, that's the point](https://github.com/serverlessworkflow/specification/blob/main/schema/functions.json#L37-L40). It's always a String for a URI (openapi, async, graphql, etc.). All of them are external references

You are speaking of the only exception that does not apply here, which are function definitions's `operation` property.
It does notes not apply becaus apply because it's not a top-level property, unlike `actions`, `retries`, etc., which ALL allow using external reference or inline object.

> Same for any other integrations we have with external standards.

No, not for top-level property. Again, the remark only applies to function definitions's `operation` property.

> A URI can be a local one (which will be in most cases). It shouldn't be an external reference all the time. This should be the recommended approach, IMHO.

Local or not, I consider references as being external (to the workflow definition), and there should not be the need to differenciate them. For proper immutable character of a workflow definition, however, a runtime cannot rely on an external (to the workflow definition) - therefore mutable - file. Especially in this case, which basically is the lifeguard mechanism of a workflow before starting them. It's part of the workflow itself, not external interopability (like functions) and as such should be defined in it. IMO, this should be the rule of thumb for any of the top level properties, with the use of `uris` being a "quality-of-life" improvement used to shorten definitions.

In Synapse, for example, we took the approach of automatically loading and explicitely embedding any and all external references, in order to striclty guarantee immutability of the version.
If you do not do such a thing, anyone could change the local (or external) file, thus altering the workflow's fundamentals (i.e. inputDataSchema) without actually touching its definition, which is, IMHO, totally unacceptable.

**Comment by ricardozanini** [comment]
> Is the concern here implementation issues? 

No, it's to make it more cohesive. I'll read carefully the other comments and answer later.

**Comment by fjtirado** [comment]
I want to remark this is not an implementation issue at all, but an aesthetic one (although the Java SDK is not currently working with inline json schema, that will need to be fixed if we want to keep it in the spec). 
I do not think a embedded Json Schema Input looks great in a yaml file. You can express an object in a yaml, but I think it will be pretty complex to render non trivial json input schema using that mechanism. 
I want also to remarks that right now, the only other thing that you can inline (If Im not missing anything) is the constants object, which is just a simple json that can be easily render in yaml. 

@cdavernas Assuming inmutability indeed simplify things. In our implementation we are loading the extenal files at runtime, storing in classpath (so we ensure are available at runtime if they were available at build time). But, you can face an scenario where the resource is not avalaible at build time (you do not have access to that URI in your built env) but you will have at runtime.

**Comment by cdavernas** [comment]
> I do not think a embedded Json Schema Input looks great in a yaml file (keeping aside that input schemas might be large). You can express an object in a yaml, but I think it will be pretty complex to render non trivial json input schema using that mechanism.

I strongly disagree, but that's a matter of taste and colors. 

Let's just say that's the path that Kubernetes chose to take, to only name that software: inline json schema. Inline because, as explained before, immutable ^^ To begin with, K8s did not even support external jsonSchema definitions!

> @cdavernas Assuming inmutability indeed simplify things

@fjtirado Not assuming, enforcing! We do not assume anything, that's actually the opposite: we enforce it: even if the user declared something as an external resource, we solve it, thus avoiding the issues your are speaking about.

> so we ensure are available at runtime if they were available at build time

My point exactly: what about build time? That sounds to me like that you are assuming those files would still be where they should, and would still the same than at declaration time, isn't it?

**Comment by fjtirado** [comment]
@cdavernas It might be possible than build environment is not equal to runtime environemnt. It is possible that a user might want to change constants values by changing an external file located in a well know place, for example
But I do not really want to discuss that, my point is that there is certain lack of coherence in allowing inlining of one standard (json input schema) not ready for YAML (as far as I know) and does not allowing inliining of another standart (OpenAPI) that support natively yaml.

**Comment by tsurdilo** [comment]
> It might be possible than build environment is not equal to runtime environemnt. It is possible that a user might want to change constants values by changing an external file located in a well know place, rather than (re)parsing the flow file, for example.

Again i think we are putting two different things into the same bucket for some reason.
Having the ability to have different function definitions on testing env and prod evn is quite useful.

To me defining the structure and required fields of your workflow data the same thing as defining the workflow definition itself, part of users business logic. As such I don't think this should differ between any environments/deployments.

I think it's pretty scary actually to allow changes to data input schema to change without a redeployment of the workflow definition itself, which kinda brings up the point if we should actually turn this around and only allow inline schemas that get deployed alongside the rest of the workflow definition core business blocks.

I think that yaml is not a problem here, at least technically it should not be.

**Comment by fjtirado** [comment]
Well, we can introduce a top level property to keep the open api definitions inlined, for example. 
I want to clarify Im not proposing that, Im just mentioning that if we allow inlining of an standard, user might ask to inlime more thing that we are currently not allowing to do so, Thats the issue

**Comment by fjtirado** [comment]
@tsurdilo ok, see your point and is a valid one. But we need to clarify that in a note somewhere.

**Comment by tsurdilo** [comment]
Yes similar to expression languages inside workflow definition. They could also be considered a "standard" but for to be able to define business logic inside workflow definition we use them.

**Comment by JBBianchi** [comment]
The argument "no inline definition because operations are URI" doesn't really make sense:
1. because operations have that weakness. If a rest endpoint is not openapi documented, it cannot be used, thus the issue #420 in which we kind of [agreed](https://github.com/serverlessworkflow/specification/issues/420#issuecomment-1059440330) on using a subset of OpenAPI "inline".
2. because there is no relation between operations and data input schema

I any case, I guess you could enforce an URI in your own runtime if you will as the spec defines it as either a string or an object. I don't see a valid justification to prevent other people from "inlining" juste because it's not "aesthetic".

**Comment by fjtirado** [comment]
@JBBianchi 
Nobody says is against inlining because operations are URIs
I was just pointing out that json input schema is the only standard we are inlining in the secped and that it does not look good when a large one is used into  a yaml file. These are two valid points (one objetive, another subjective)
I will fix the SDK to be aligned with spec (The Java SDK fails to parse the document when using inline data input) 
Again, we do not have any implementation issue using inline (besides the fact that is blocked by the SDK)

**Comment by fjtirado** [comment]
> Each workflow will probably have its own input definition, it doesn't make that much sense to have a reference to it. For an operation, it does, a lot of workflows can use the same rest end point.

Hmmm, that Im not sure that will be true. Actually, one strong argument for not inlining  by default is that it will encourage reuse of schemas. I can certainly see the same object being used in a lot of different flows (employee data, arrays of numbers...)

**Comment by ricardozanini** [comment]
> That's true but then I'd rather rely on JSON Schema's `$ref`.

JSON Schema's `$ref` inside an inline workflow definition will be awkward, IMO.

I can see the reasoning of "let's inline, the data structure definition for this workflow is clear." This is fine.

The problem is this mixture of standards in the workflow definition. But, again, from an implementation pov, it doesn't matter. We can assume the `ref` is in the same or any other path as specified there and build from there.

Well, I believe that we are going nowhere in this discussion. And I see that doesn't worth adding to the spec that implementations can either support both. Implementations can document how they handle JSON Schemas.

If you work with external JSON Schemas, as @fjtirado mentioned, you can easily reuse them (that's likely the case in a production env). Additionally, you can use external tools to manage these files for you, and you just add the reference in your workflow definition.

Part of the process/parsing of the workflow with `refs` inlined in the input schema, should or should not load them? This is also a valid concern, also any kind of JSON Schema processing should happen on our side as well.

**Comment by JBBianchi** [comment]
Just saying inline is fine, URI is fine, both should be supported. 

As for `$ref`, it's just that I kind of agree with @fjtirado in the sense where common entities' schemas will probably be used and, therefore, accessing URI/external files makes sense. So, you could have `"dataInputSchema": "somewhere/employee.json"`, that is a valid idea. But, I'm also thinking that each workflow will probably start with it's own "command", it's own "payload", with named properties that can vary eventhough they have common types. For instance `"dataInputSchema": { "type": "object", "properties": { "employee": { "$ref": "somewhere/employee.json" }, "action": { "type": "string" } } }` and you will want each workflow to have its own to prevent side effects if you change that "command"/"payload", however you call it.

**Comment by tsurdilo** [comment]
Can we mark this as resolved or is there a need for further discussions? Please let me know.

**Comment by fjtirado** [comment]
@tsurdilo discussion is resolved ;). Spec will support both reference and inline
